<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Shiny App to Explore RNA-Seq Analysis • carnation</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Shiny App to Explore RNA-Seq Analysis">
<meta name="description" content="Interactive Shiny dashboard app that can be used to explore RNA-Seq analysis results including differential expression (DE), functional enrichment and pattern analysis. Several visualizations are implemented to provide a wide-ranging view of data sets. For DE analysis, we provide PCA plot, MA plot, Upset plot &amp; heatmaps, in addition to a highly customizable gene plot. Seven different visualizations are available for functional enrichment analysis, and we also support gene pattern analysis. In addition, the app provides a platform to manage multiple projects and user groups that can be run on a central server.">
<meta property="og:description" content="Interactive Shiny dashboard app that can be used to explore RNA-Seq analysis results including differential expression (DE), functional enrichment and pattern analysis. Several visualizations are implemented to provide a wide-ranging view of data sets. For DE analysis, we provide PCA plot, MA plot, Upset plot &amp; heatmaps, in addition to a highly customizable gene plot. Seven different visualizations are available for functional enrichment analysis, and we also support gene pattern analysis. In addition, the app provides a platform to manage multiple projects and user groups that can be run on a central server.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">carnation</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/carnation.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="carnation">Carnation<a class="anchor" aria-label="anchor" href="#carnation"></a>
</h1></div>
<p>Interactive shiny app for deeply exploring bulk RNA-Seq data, incorporating functional enrichment and pattern analysis.</p>
<ul>
<li>Assess differential expression analysis results using interactive PCA, UpSet plots, heatmaps and a highly customizable gene plot.</li>
<li>Dive into functional enrichment analysis with the help of fuzzy search across genes or pathway descriptions.</li>
<li>Easily track genes of interest across the app using the novel “Gene scratchpad”.</li>
<li>Data access interface with an inbuilt authentication layer can be used in server mode to share results with others or to organize RNA-Seq projects on your local machine.</li>
</ul>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<div class="section level2">
<h2 id="condamamba-recommended">conda/mamba (recommended)<a class="anchor" aria-label="anchor" href="#condamamba-recommended"></a>
</h2>
<p>You can install carnation by creating a new conda environment containing all dependencies as listed in the file <code>requirements-pinned.yaml</code>. This is <em>recommended</em>.</p>
<p><strong>WARNING:</strong> Make sure you build the conda environment <em>outside</em> the carnation directory or the build step will fail.</p>
<pre><code>cd .. &amp;&amp; mamba env create -p env --file carnation/requirements-pinned.yaml</code></pre>
<p>Next, activate the environment and install using <code>R CMD build</code> and <code>install.packages</code>:</p>
<pre><code>conda activate ./env
R CMD build carnation/
Rscript -e "install.packages('carnation_1.0.tar.gz', repos=NULL)"</code></pre>
</div>
<div class="section level2">
<h2 id="devtools">devtools<a class="anchor" aria-label="anchor" href="#devtools"></a>
</h2>
<p>You can also install carnation using the <code>devtools</code> package. Here, we use <code>devtools</code> to figure out the dependencies directly, mimicking a direct R installation.</p>
<p>First, install <code>devtools</code> in an existing R installation, e.g. in RStudio.</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'devtools'</span><span class="op">)</span></span></code></pre>
<p>If conda/mamba is installed, you could also do this with a conda environment that contains only <code>devtools</code>, e.g.</p>
<pre><code>conda create -p env r-devtools
conda activate ./env</code></pre>
<p>Next, open R, set repositories to get both CRAN and bioconductor packages and run <code>devtools::install_github</code>.</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/setRepositories.html" class="external-link">setRepositories</a></span><span class="op">(</span>ind<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">4</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">'NICHD-BSPC/carnation'</span><span class="op">)</span></span></code></pre>
<p>Note: This may be a time-consuming step, especially if working with a fresh R installation, as a large number of dependencies will be installed.</p>
</div>
</div>
<div class="section level1">
<h1 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h1>
<div class="section level2">
<h2 id="data-setup">Data setup<a class="anchor" aria-label="anchor" href="#data-setup"></a>
</h2>
<p>Organize your data in a directory structure that carnation can use, e.g. in <code>/carnation/data</code>. For example, let’s say you have two projects <em>project1</em> and <em>project2</em>.</p>
<ul>
<li>
<em>project1</em> has 3 analyses, <em>main</em>, <em>subset</em> and <em>main-nooutlier</em>
</li>
<li>
<em>project2</em> has 1 analysis, <em>default</em>
</li>
</ul>
<p>Then your carnation data directory can look like this:</p>
<pre><code>/carnation/data/
  ├─ project1
  │  ├─ main.rds
  │  ├─ subset.rds
  │  └─ main-nooutlier.rds
  │
  └─ project2
     └─ default.rds</code></pre>
<p>The first time you run carnation, you will be asked to choose a data area. If you have set up your data as above, type in the location <code>/carnation/data</code> in the data area input.</p>
<p>If everything worked as expected, then the <em>Available projects</em> menu should be populated with two options <em>project1</em> and <em>project2</em>.</p>
<ul>
<li>If <em>project1</em> is selected, <em>Available assays</em> will show three options, <em>main</em>, <em>subset</em> and <em>main-nooutlier</em>.</li>
<li>If <em>project2</em> is selected, there will be one available assay - <em>default</em>.</li>
</ul>
</div>
<div class="section level2">
<h2 id="first-run">First run<a class="anchor" aria-label="anchor" href="#first-run"></a>
</h2>
<p>First, load the <code>carnation</code> library.</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nichd-bspc.github.io/carnation/">carnation</a></span><span class="op">)</span></span></code></pre>
<p><code>carnation</code> uses the <code>kaleido</code> python module to allow interactive plots to be saved to PDF. To enable this functionality, you need to run:</p>
<pre><code><span><span class="fu"><a href="reference/install_carnation.html">install_carnation</a></span><span class="op">(</span><span class="op">)</span></span></code></pre>
<p>By default, this installs the python modules <code>plotly</code> and <code>kaleido</code> into a python virtual environment named <code>r-carnation</code>.</p>
<p>Finally, run the main app function.</p>
<pre><code><span><span class="fu"><a href="reference/run_carnation.html">run_carnation</a></span><span class="op">(</span><span class="op">)</span></span></code></pre>
<p>This will open up carnation in a browser window. Note that, if you’re running carnation from a remote ssh server, you might need to add some additional options. For example, if you’re running carnation on a remote server that is forwarding port <code>12345</code> via ssh tunneling:</p>
<pre><code><span><span class="fu"><a href="reference/run_carnation.html">run_carnation</a></span><span class="op">(</span>options<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>port<span class="op">=</span><span class="fl">12345</span>, launch.browser<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>This will start carnation on port <code>12345</code>.</p>
<p>Now point your web browser to the URL: <code>http://127.0.0.1:12345</code>. Since, this is your first run, carnation will ask you where your data is via a modal dialog.</p>
</div>
</div>
<div class="section level1">
<h1 id="server-mode">Server mode<a class="anchor" aria-label="anchor" href="#server-mode"></a>
</h1>
<p>Carnation supports running in server mode using <code>shinymanager</code> to add an authentication layer. For this, you need to first create a local sqlite database with user details (for more details see <a href="https://datastorm-open.github.io/shinymanager/" class="external-link">shinymanager docs</a>).</p>
<pre><code><span><span class="co"># create data frame with user details</span></span>
<span><span class="va">credentials</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  user <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'shinymanager'</span><span class="op">)</span>,</span>
<span>  password <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'12345'</span><span class="op">)</span>,</span>
<span>  <span class="co"># password will automatically be hashed</span></span>
<span>  admin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Init the database</span></span>
<span><span class="fu">shinymanager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/shinymanager/man/create_db.html" class="external-link">create_db</a></span><span class="op">(</span></span>
<span>  credentials_data <span class="op">=</span> <span class="va">credentials</span>,</span>
<span>  sqlite_path <span class="op">=</span> <span class="st">'credentials.sqlite'</span>, <span class="co"># will be created</span></span>
<span>  passphrase <span class="op">=</span> <span class="st">'admin_passphrase'</span></span>
<span><span class="op">)</span></span></code></pre>
<p>This will create an sqlite file with an admin user ‘shinymanager’. Next, run carnation pointing to the credentials you just created.</p>
<pre><code><span><span class="fu"><a href="reference/run_carnation.html">run_carnation</a></span><span class="op">(</span>credentials<span class="op">=</span><span class="st">'credentials.sqlite'</span>, passphrase<span class="op">=</span><span class="st">'admin_passphrase'</span><span class="op">)</span></span></code></pre>
<p>Now, you will see a login page asking you to authenticate before you can access carnation.</p>
</div>

  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing carnation</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Apratim Mitra <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-3279-0054" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Apratim Mitra.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
