<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>carnation - airway tutorial • carnation</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="carnation - airway tutorial">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">carnation</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/carnation.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/NICHD-BSPC/carnation/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>carnation - airway tutorial</h1>
                        <h4 data-toc-skip class="author">Apratim
Mitra</h4>
            
            <h4 data-toc-skip class="date">12/23/2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/NICHD-BSPC/carnation/blob/main/vignettes/carnation.Rmd" class="external-link"><code>vignettes/carnation.Rmd</code></a></small>
      <div class="d-none name"><code>carnation.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Carnation is an interactive Shiny dashboard that simplifies and
      transforms complex bulk RNA-Seq data using insightful
      visualizations and numerous interactive features. Designed for
      both computational and experimental biologists, Carnation makes
      exploring differential expression analysis, functional enrichment,
      and pattern analysis intuitive and exciting. In this vignette, we
      will describe some of its functionality using the
      <code>airway</code> dataset. Carnation package version: 0.99.3</p>
    </div>
    
<div class="section level3">
<h3 id="load-libraries-airway-dataset">Load libraries &amp; airway dataset<a class="anchor" aria-label="anchor" href="#load-libraries-airway-dataset"></a>
</h3>
<p>First check for and load some libraries that we will need for this
tutorial.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install optional packages if not present</span></span>
<span><span class="va">pkgs_to_check</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'airway'</span>, <span class="st">'org.Hs.eg.db'</span>, <span class="st">'clusterProfiler'</span>, <span class="st">'DEGreport'</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">pkg</span> <span class="kw">in</span> <span class="va">pkgs_to_check</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/setRepositories.html" class="external-link">setRepositories</a></span><span class="op">(</span>ind<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">4</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="va">pkg</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="va">pkg</span>, repos<span class="op">=</span><span class="st">'http://cran.us.r-project.org'</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/federicomarini/GeneTonic" class="external-link">GeneTonic</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/contribution-knowledge-mining/" class="external-link">clusterProfiler</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://lpantano.github.io/DEGreport/" class="external-link">DEGreport</a></span><span class="op">)</span></span></code></pre></div>
<p>We will be using the ‘airway’ dataset. First, we load this
dataset.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">'airway'</span><span class="op">)</span></span></code></pre></div>
<p>Next, we extract the counts matrix and and metadata.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span><span class="va">cdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span></code></pre></div>
<p>Now let’s see what these look like.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 63677     8</span></span></code></pre></div>
<p>So, <code>mat</code> is a matrix with 64102 rows and 8 columns. Each
row corresponds to a single gene and each column corresponds to a single
sample. As you will notice, the rownames of <code>mat</code> contains
gene IDs and column names have the sample IDs.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 SRR1039508 SRR1039509 SRR1039512 SRR1039513 SRR1039516</span></span>
<span><span class="co">#&gt; ENSG00000000003        679        448        873        408       1138</span></span>
<span><span class="co">#&gt; ENSG00000000005          0          0          0          0          0</span></span>
<span><span class="co">#&gt; ENSG00000000419        467        515        621        365        587</span></span>
<span><span class="co">#&gt; ENSG00000000457        260        211        263        164        245</span></span>
<span><span class="co">#&gt; ENSG00000000460         60         55         40         35         78</span></span>
<span><span class="co">#&gt; ENSG00000000938          0          0          2          0          1</span></span>
<span><span class="co">#&gt;                 SRR1039517 SRR1039520 SRR1039521</span></span>
<span><span class="co">#&gt; ENSG00000000003       1047        770        572</span></span>
<span><span class="co">#&gt; ENSG00000000005          0          0          0</span></span>
<span><span class="co">#&gt; ENSG00000000419        799        417        508</span></span>
<span><span class="co">#&gt; ENSG00000000457        331        233        229</span></span>
<span><span class="co">#&gt; ENSG00000000460         63         76         60</span></span>
<span><span class="co">#&gt; ENSG00000000938          0          0          0</span></span></code></pre></div>
<p><code>cdata</code> contains the sample metadata. There is a lot of
information here, but notice the <code>cell</code> and <code>dex</code>
columns, as we will be using this for the differential expression
analysis later.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cdata</span></span>
<span><span class="co">#&gt; DataFrame with 8 rows and 9 columns</span></span>
<span><span class="co">#&gt;            SampleName     cell      dex    albut        Run avgLength</span></span>
<span><span class="co">#&gt;              &lt;factor&gt; &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;   &lt;factor&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt; SRR1039508 GSM1275862  N61311     untrt    untrt SRR1039508       126</span></span>
<span><span class="co">#&gt; SRR1039509 GSM1275863  N61311     trt      untrt SRR1039509       126</span></span>
<span><span class="co">#&gt; SRR1039512 GSM1275866  N052611    untrt    untrt SRR1039512       126</span></span>
<span><span class="co">#&gt; SRR1039513 GSM1275867  N052611    trt      untrt SRR1039513        87</span></span>
<span><span class="co">#&gt; SRR1039516 GSM1275870  N080611    untrt    untrt SRR1039516       120</span></span>
<span><span class="co">#&gt; SRR1039517 GSM1275871  N080611    trt      untrt SRR1039517       126</span></span>
<span><span class="co">#&gt; SRR1039520 GSM1275874  N061011    untrt    untrt SRR1039520       101</span></span>
<span><span class="co">#&gt; SRR1039521 GSM1275875  N061011    trt      untrt SRR1039521        98</span></span>
<span><span class="co">#&gt;            Experiment    Sample    BioSample</span></span>
<span><span class="co">#&gt;              &lt;factor&gt;  &lt;factor&gt;     &lt;factor&gt;</span></span>
<span><span class="co">#&gt; SRR1039508  SRX384345 SRS508568 SAMN02422669</span></span>
<span><span class="co">#&gt; SRR1039509  SRX384346 SRS508567 SAMN02422675</span></span>
<span><span class="co">#&gt; SRR1039512  SRX384349 SRS508571 SAMN02422678</span></span>
<span><span class="co">#&gt; SRR1039513  SRX384350 SRS508572 SAMN02422670</span></span>
<span><span class="co">#&gt; SRR1039516  SRX384353 SRS508575 SAMN02422682</span></span>
<span><span class="co">#&gt; SRR1039517  SRX384354 SRS508576 SAMN02422673</span></span>
<span><span class="co">#&gt; SRR1039520  SRX384357 SRS508579 SAMN02422683</span></span>
<span><span class="co">#&gt; SRR1039521  SRX384358 SRS508580 SAMN02422677</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="get-more-gene-annotation">Get more gene annotation<a class="anchor" aria-label="anchor" href="#get-more-gene-annotation"></a>
</h3>
<p>The gene IDs that come with the dataset are from ENSEMBL and are not
human-readable. So, next we will extract gene symbols and
<code>ENTREZID</code> for these genes from the <code>org.Hs.eg.db</code>
package.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keytypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'SYMBOL'</span><span class="op">=</span><span class="st">'SYMBOL'</span>, <span class="st">'ENTREZID'</span><span class="op">=</span><span class="st">'ENTREZID'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">anno_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">'cbind'</span>,</span>
<span>             <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">keytypes</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>               <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>,</span>
<span>                 column<span class="op">=</span><span class="va">x</span>,</span>
<span>                 keys<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span>,</span>
<span>                 keytype<span class="op">=</span><span class="st">'ENSEMBL'</span><span class="op">)</span></span>
<span>               <span class="op">}</span><span class="op">)</span></span>
<span>             <span class="op">)</span></span>
<span></span>
<span><span class="co"># convert to data frame</span></span>
<span><span class="va">anno_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">anno_df</span><span class="op">)</span></span></code></pre></div>
<p>Now, we have human readable gene names in the <code>SYMBOL</code>
column and Entrez IDs in the <code>ENTREZ</code> column.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">anno_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 SYMBOL ENTREZID</span></span>
<span><span class="co">#&gt; ENSG00000000003 TSPAN6     7105</span></span>
<span><span class="co">#&gt; ENSG00000000005   TNMD    64102</span></span>
<span><span class="co">#&gt; ENSG00000000419   DPM1     8813</span></span>
<span><span class="co">#&gt; ENSG00000000457  SCYL3    57147</span></span>
<span><span class="co">#&gt; ENSG00000000460  FIRRM    55732</span></span>
<span><span class="co">#&gt; ENSG00000000938    FGR     2268</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-deseqdataset">Create DESeqDataSet<a class="anchor" aria-label="anchor" href="#create-deseqdataset"></a>
</h3>
<p>Next, we create a new <code>DESeqDataSet</code> using
<code>mat</code> and <code>cdata</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSetFromMatrix</a></span><span class="op">(</span><span class="va">mat</span>,</span>
<span>                              colData<span class="op">=</span><span class="va">cdata</span>,</span>
<span>                              design<span class="op">=</span><span class="op">~</span><span class="va">cell</span> <span class="op">+</span> <span class="va">dex</span><span class="op">)</span></span></code></pre></div>
<p>Let’s check to make sure that everything looks okay:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span></span>
<span><span class="co">#&gt; class: DESeqDataSet </span></span>
<span><span class="co">#&gt; dim: 63677 8 </span></span>
<span><span class="co">#&gt; metadata(1): version</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(63677): ENSG00000000003 ENSG00000000005 ... ENSG00000273492</span></span>
<span><span class="co">#&gt;   ENSG00000273493</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521</span></span>
<span><span class="co">#&gt; colData names(9): SampleName cell ... Sample BioSample</span></span></code></pre></div>
<p>Then we save <code>dds</code> in a list.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>main<span class="op">=</span><span class="va">dds</span><span class="op">)</span></span></code></pre></div>
<p>We also normalize the counts data and save it in a list.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rld_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">dds_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/varianceStabilizingTransformation.html" class="external-link">varianceStabilizingTransformation</a></span><span class="op">(</span><span class="va">x</span>, blind<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-differential-expression-analysis">Run differential expression analysis<a class="anchor" aria-label="anchor" href="#run-differential-expression-analysis"></a>
</h3>
<p>Now, the object <code>dds</code> is ready for differential expression
(DE) analysis.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="co">#&gt; estimating size factors</span></span>
<span><span class="co">#&gt; estimating dispersions</span></span>
<span><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span><span class="co">#&gt; mean-dispersion relationship</span></span>
<span><span class="co">#&gt; final dispersion estimates</span></span>
<span><span class="co">#&gt; fitting model and testing</span></span></code></pre></div>
<p>Since, we used <code>design=~cell + dex</code> while creating
<code>dds</code>, the above step will automatically calculate some
comparisons.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">resultsNames</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Intercept"               "cell_N061011_vs_N052611"</span></span>
<span><span class="co">#&gt; [3] "cell_N080611_vs_N052611" "cell_N61311_vs_N052611" </span></span>
<span><span class="co">#&gt; [5] "dex_untrt_vs_trt"</span></span></code></pre></div>
<p>The last comparison <code>dex_untrt_vs_trt</code> contains the effect
of the <code>dex</code> treatment, while the other comparisons compare
different cell lines. Next, we will extract two of these results and run
<code>lfcShrink</code> on them.</p>
<p>For the <code>cell</code> comparison, we choose a precomputed results
using the <code>coef</code> parameter.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_comparison</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/lfcShrink.html" class="external-link">lfcShrink</a></span><span class="op">(</span><span class="va">dds</span>,</span>
<span>                             coef<span class="op">=</span><span class="st">'cell_N61311_vs_N052611'</span>,</span>
<span>                             type<span class="op">=</span><span class="st">'normal'</span><span class="op">)</span></span></code></pre></div>
<p>For the <code>dex</code> comparison, we use <code>contrast</code> to
specify the direction of the comparison, since we want to use
<code>untrt</code> as control.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dex_comparison</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/lfcShrink.html" class="external-link">lfcShrink</a></span><span class="op">(</span><span class="va">dds</span>,</span>
<span>                            contrast<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'dex'</span>, <span class="st">'trt'</span>, <span class="st">'untrt'</span><span class="op">)</span>,</span>
<span>                            type<span class="op">=</span><span class="st">'normal'</span><span class="op">)</span></span></code></pre></div>
<p>Now, each of these comparisons, contain the DE analysis results. For
example,</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dex_comparison</span><span class="op">)</span></span>
<span><span class="co">#&gt; log2 fold change (MAP): dex trt vs untrt </span></span>
<span><span class="co">#&gt; Wald test p-value: dex trt vs untrt </span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 6 columns</span></span>
<span><span class="co">#&gt;                   baseMean log2FoldChange     lfcSE      stat      pvalue</span></span>
<span><span class="co">#&gt;                  &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSG00000000003 708.602170     -0.3741527 0.0988429 -3.787752 0.000152016</span></span>
<span><span class="co">#&gt; ENSG00000000005   0.000000             NA        NA        NA          NA</span></span>
<span><span class="co">#&gt; ENSG00000000419 520.297901      0.2020620 0.1097395  1.842943 0.065337292</span></span>
<span><span class="co">#&gt; ENSG00000000457 237.163037      0.0361672 0.1383377  0.264356 0.791505742</span></span>
<span><span class="co">#&gt; ENSG00000000460  57.932633     -0.0844567 0.2498904 -0.307054 0.758801924</span></span>
<span><span class="co">#&gt; ENSG00000000938   0.318098     -0.0841390 0.1513343 -0.393793 0.693733530</span></span>
<span><span class="co">#&gt;                       padj</span></span>
<span><span class="co">#&gt;                  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSG00000000003 0.00128292</span></span>
<span><span class="co">#&gt; ENSG00000000005         NA</span></span>
<span><span class="co">#&gt; ENSG00000000419 0.19646985</span></span>
<span><span class="co">#&gt; ENSG00000000457 0.91141962</span></span>
<span><span class="co">#&gt; ENSG00000000460 0.89500478</span></span>
<span><span class="co">#&gt; ENSG00000000938         NA</span></span></code></pre></div>
<p>Then, we save these results in a special nested list that
<code>carnation</code> will use. Here,</p>
<ul>
<li>
<code>res</code> contains the actual DE analysis results</li>
<li>
<code>dds</code> contains the name of the DESeqDataSet used for the
DE analysis. These values should map to <code>dds_list</code> names</li>
<li>
<code>label</code> is a description of the comparison</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        dex_trt_vs_untrt<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>            res<span class="op">=</span><span class="va">dex_comparison</span>,</span>
<span>            dds<span class="op">=</span><span class="st">'main'</span>,</span>
<span>            label<span class="op">=</span><span class="st">'dex, treated vs untreated'</span><span class="op">)</span>,</span>
<span>        cell_N61311_vs_N052611<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>            res<span class="op">=</span><span class="va">cell_comparison</span>,</span>
<span>            dds<span class="op">=</span><span class="st">'main'</span>,</span>
<span>            label<span class="op">=</span><span class="st">'cell, N61311 vs N052611'</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span></code></pre></div>
<p>Finally, we add <code>SYMBOL</code> and <code>ENTREZID</code> columns
to the DE results from the <code>anno_df</code> data frame.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">res_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>              <span class="co"># save the rownames as a new 'gene' column</span></span>
<span>              <span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">[[</span> <span class="st">'gene'</span> <span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">)</span></span>
<span></span>
<span>              <span class="co"># add 'SYMBOL' and 'ENTREZID' columns</span></span>
<span>              <span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">[[</span> <span class="st">'SYMBOL'</span> <span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">anno_df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">)</span>, <span class="st">'SYMBOL'</span><span class="op">]</span></span>
<span>              <span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">[[</span> <span class="st">'ENTREZID'</span> <span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">anno_df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">)</span>, <span class="st">'ENTREZID'</span><span class="op">]</span></span>
<span></span>
<span>              <span class="va">x</span></span>
<span>            <span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-functional-enrichment-results-optional">Add functional enrichment results (optional)<a class="anchor" aria-label="anchor" href="#add-functional-enrichment-results-optional"></a>
</h3>
<p>Now we run functional enrichment on the DE genes from the two
comparisons. For this, we first set significance thresholds and then
extract the DE genes and save as a list.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># padj cutoff</span></span>
<span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.01</span></span>
<span></span>
<span><span class="co"># log2FoldChange threshold; 1 == 2x difference</span></span>
<span><span class="va">lfc_threshold</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span></span>
<span><span class="co"># list to save DE genes</span></span>
<span><span class="va">de.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">res_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>              <span class="co"># changed genes</span></span>
<span>              <span class="va">idx</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="va">alpha</span> <span class="op">&amp;</span></span>
<span>                     <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">$</span><span class="va">padj</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>                     <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">lfc_threshold</span></span>
<span></span>
<span>              <span class="co"># return DE genes as a dataframe</span></span>
<span>              <span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">[</span><span class="va">idx</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'gene'</span>, <span class="st">'ENTREZID'</span><span class="op">)</span><span class="op">]</span></span>
<span>            <span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>Next, we run functional enrichment and save the results in a list
called <code>enrich_list</code>. We also save a converted list called
<code>genetonic</code> which carnation uses for several plots from the
<code>GeneTonic</code> package.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># list to save functional enrichment results</span></span>
<span><span class="va">enrich_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># list to save a converted object for GeneTonic plots</span></span>
<span><span class="va">genetonic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">comp</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res_list</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">go.res</span> <span class="op">&lt;-</span> <span class="fu">clusterProfiler</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html" class="external-link">enrichGO</a></span><span class="op">(</span></span>
<span>            gene<span class="op">=</span><span class="va">de.genes</span><span class="op">[[</span><span class="va">comp</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">'ENTREZID'</span><span class="op">]</span><span class="op">]</span>,</span>
<span>            keyType<span class="op">=</span><span class="st">'ENTREZID'</span>,</span>
<span>            OrgDb<span class="op">=</span><span class="va">org.Hs.eg.db</span>,</span>
<span>            ont<span class="op">=</span><span class="st">'BP'</span>,</span>
<span>            pvalueCutoff<span class="op">=</span><span class="fl">1</span>, qvalueCutoff<span class="op">=</span><span class="fl">1</span>,</span>
<span>            readable <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">enrich_list</span><span class="op">[[</span> <span class="va">comp</span> <span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                               res<span class="op">=</span><span class="va">comp</span>,</span>
<span>                               changed<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span> BP<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">go.res</span><span class="op">)</span> <span class="op">)</span></span>
<span>                             <span class="op">)</span></span>
<span></span>
<span>    <span class="va">genetonic</span><span class="op">[[</span> <span class="va">comp</span> <span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                             res<span class="op">=</span><span class="va">comp</span>,</span>
<span>                             changed<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                               BP<span class="op">=</span><span class="fu">carnation</span><span class="fu">::</span><span class="fu"><a href="../reference/enrich_to_genetonic.html">enrich_to_genetonic</a></span><span class="op">(</span><span class="va">go.res</span>, <span class="va">res_list</span><span class="op">[[</span><span class="va">comp</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">res</span><span class="op">)</span></span>
<span>                             <span class="op">)</span></span>
<span>                           <span class="op">)</span></span>
<span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; Found 4643 gene sets in `enrichResult` object, of which 4643 are significant.</span></span>
<span><span class="co">#&gt; Converting for usage in GeneTonic...</span></span>
<span><span class="co">#&gt; Found 2405 gene sets in `enrichResult` object, of which 2405 are significant.</span></span>
<span><span class="co">#&gt; Converting for usage in GeneTonic...</span></span></code></pre></div>
<p><code>enrich_list</code> is a nested list where:</p>
<ul>
<li>The top-level names are unique keys/identifiers</li>
<li>The second level corresponds to direction of change,
e.g. <code>up</code>, <code>down</code> or <code>changed</code>. This
level also contains a special entry <code>res</code> which maps to
<code>res_list</code> names, as a way to record where the DE results
came from.</li>
<li>The third level corresponds to the ontology, e.g. <code>BP</code>
(GO Biological Process).</li>
</ul>
<p>Here, we are just using changed genes and the GO Biological Process
(BP) ontology, so <code>enrich_list</code> looks like:</p>
<pre><code>enrich_list
  ├─ dex_trt_vs_untrt
  │    ├─ res = 'des_trt_vs_untrt'    &lt;--- comparison used to get FE results
  │    └─ changed
  │         └─ BP                     &lt;--- functional enrichment results
  │
  └─ cell_N61311_vs_N052611
       ├─ res = 'cell_N61311_vs_N052611'    &lt;--- comparison used to get FE results
       └─ changed
            └─ BP                           &lt;--- functional enrichment results
</code></pre>
<p><code>genetonic</code> mirrors the same structure.</p>
</div>
<div class="section level3">
<h3 id="add-pattern-analysis-optional">Add pattern analysis (optional)<a class="anchor" aria-label="anchor" href="#add-pattern-analysis-optional"></a>
</h3>
<p>Finally, we add some pattern analysis for the
<code>dex_trt_vs_untrt</code> comparison using the
<code>DEGreport</code> package. First, we extract normalized data for
the 755 DE genes from this comparison.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract normalized data &amp; metadata</span></span>
<span><span class="va">ma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rld_list</span><span class="op">[[</span><span class="st">'main'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">colData.i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rld_list</span><span class="op">[[</span><span class="st">'main'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># only keep data from DE genes</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ma</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">de.genes</span><span class="op">[[</span><span class="st">'dex_trt_vs_untrt'</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">'gene'</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">ma.i</span> <span class="op">&lt;-</span> <span class="va">ma</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># remove any genes with 0 variance</span></span>
<span><span class="va">ma.i</span> <span class="op">&lt;-</span> <span class="va">ma.i</span><span class="op">[</span><span class="fu">rowVars</span><span class="op">(</span><span class="va">ma.i</span><span class="op">)</span> <span class="op">!=</span> <span class="fl">0</span>, <span class="op">]</span></span></code></pre></div>
<p>Then, we run the pattern analysis, using <code>cell</code> as the
<em>time</em> variable and <code>dex</code> as the <em>color</em>
variable.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">DEGreport</span><span class="fu">::</span><span class="fu"><a href="http://lpantano.github.io/DEGreport/reference/degPatterns.html" class="external-link">degPatterns</a></span><span class="op">(</span></span>
<span>        <span class="va">ma.i</span>,</span>
<span>        <span class="va">colData.i</span>,</span>
<span></span>
<span>        time<span class="op">=</span><span class="st">'cell'</span>,</span>
<span>        col<span class="op">=</span><span class="st">'dex'</span>,</span>
<span></span>
<span>        <span class="co"># NOTE: reduce and merge cutoff----------------------------------------</span></span>
<span>        <span class="co">#   Reduce will merge clusters that are similar; similarity determined</span></span>
<span>        <span class="co">#   by cutoff</span></span>
<span>        reduce<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span></span>
<span>        <span class="co"># NOTE: set minimum cluster size to 1</span></span>
<span>        minc<span class="op">=</span><span class="fl">1</span>,</span>
<span>        plot<span class="op">=</span><span class="cn">FALSE</span></span>
<span>        <span class="op">)</span></span>
<span><span class="co">#&gt; Working with 754 genes.</span></span>
<span><span class="co">#&gt; Working with 650 genes after filtering: minc &gt; 1</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(merge)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(merge)`</span></span></code></pre></div>
<p>Next, we extract the <code>normalized</code> slot from this object
and save as a list.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract normalized slot and add symbol column</span></span>
<span><span class="va">p_norm</span> <span class="op">&lt;-</span> <span class="va">p</span><span class="op">$</span><span class="va">normalized</span></span>
<span><span class="va">p_norm</span><span class="op">[[</span> <span class="st">'SYMBOL'</span> <span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">anno_df</span><span class="op">[</span><span class="va">p_norm</span><span class="op">[[</span><span class="st">'genes'</span><span class="op">]</span><span class="op">]</span>, <span class="st">'SYMBOL'</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># save pattern analysis results</span></span>
<span><span class="va">degpatterns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>dex_by_cell<span class="op">=</span><span class="va">p_norm</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="compose-carnation-object">Compose carnation object<a class="anchor" aria-label="anchor" href="#compose-carnation-object"></a>
</h3>
<p>Now we have all the pieces to build the carnation object.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>res_list<span class="op">=</span><span class="va">res_list</span>,</span>
<span>                 dds_list<span class="op">=</span><span class="va">dds_list</span>,</span>
<span>                 rld_list<span class="op">=</span><span class="va">rld_list</span>,</span>
<span>                 enrich_list<span class="op">=</span><span class="va">enrich_list</span>,</span>
<span>                 genetonic<span class="op">=</span><span class="va">genetonic</span>,</span>
<span>                 degpatterns_list<span class="op">=</span><span class="va">degpatterns</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">combined</span>, <span class="st">'carnation_vignette.rds'</span>, compress<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data-organization">Data Organization<a class="anchor" aria-label="anchor" href="#data-organization"></a>
</h3>
<p>Organize your data in a directory structure that Carnation can easily
navigate, e.g. in a folder <code>carnation/data</code> in your home
directory. Here, we make a subfolder <code>airway</code> in this folder
and copy the <code>RDS</code> file we created above.</p>
<pre><code>~/carnation/data/
  └─ airway
     └─ carnation-vignette.rds</code></pre>
<p>Now, we’re ready for the first run.</p>
</div>
<div class="section level3">
<h3 id="first-run">First Run<a class="anchor" aria-label="anchor" href="#first-run"></a>
</h3>
<p>Load the carnation package.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nichd-bspc.github.io/carnation/">carnation</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'carnation'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:GeneTonic':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     gs_radar</span></span></code></pre></div>
<p>Carnation allows you to download interactive plots as PDF. To use
this functionality you need to install some required Python
dependencies:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/install_carnation.html">install_carnation</a></span><span class="op">(</span><span class="op">)</span>  <span class="co"># Installs plotly and kaleido for PDF export</span></span></code></pre></div>
<p>Now, run the app:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/run_carnation.html">run_carnation</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>To run on a fixed port, e.g. when using remote servers with SSH port
forwarding, specify <code>port</code> within a list of options.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/run_carnation.html">run_carnation</a></span><span class="op">(</span>options<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>port<span class="op">=</span><span class="fl">12345</span>, launch.browser<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Then access Carnation by opening the URL:
<code>http://127.0.0.1:12345</code></p>
<div class="section level4">
<h4 id="initial-setup">Initial setup<a class="anchor" aria-label="anchor" href="#initial-setup"></a>
</h4>
<ul>
<li>
<p>The first time you run carnation, you will be asked to specify
where your data is located with a modal dialog.</p>
<p>Enter the folder where you saved the RDS file,
<code>~/carnation/data</code>, then click ‘OK’.</p>
<p><img src="images/01_first_run.png" width="600"></p>
</li>
<li>
<p>Now carnation will automatically refresh and you will see
<code>data/airway</code> in the “Available Projects” menu, and
<code>carnation_vignette</code> in the “Available Datasets”.</p>
<p>Since this is the only dataset we have, these options are
automatically selected. Now click ‘Go’ to load the data.</p>
<p><img src="images/02_pre_load.png" width="600"></p>
<p>Now carnation will load the airway data and switch to the “Summary”
tab.</p>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="general-layout">General layout<a class="anchor" aria-label="anchor" href="#general-layout"></a>
</h4>
<p>The interface of Carnation consists of a main central panel for
content, a sidebar containing settings and ‘Gene scratchpad’ buttons and
context-specific help buttons.</p>
<p><img src="images/03_layout.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="feature-overview">Feature overview<a class="anchor" aria-label="anchor" href="#feature-overview"></a>
</h3>
<p>Carnation is organized into three major sections each with its own
set of subtabs to organize different portions of the RNA-Seq
analysis.</p>
<div class="section level4">
<h4 id="de-analysis">DE analysis<a class="anchor" aria-label="anchor" href="#de-analysis"></a>
</h4>
<p>Here you can analyze differential expression through different tables
and visualizations:</p>
<ul>
<li>
<p><em>Summary</em>: Get a quick overview of your differential
expression results</p>
<p><img src="images/04_data_loaded.png" width="600"></p>
</li>
<li>
<p><em>Metadata</em>: Explore sample metadata and experimental
design. You can also add metadata columns.</p>
<p><img src="images/05_metadata.png" width="600"></p>
</li>
<li>
<p><em>PCA Plot</em>: Visualize sample relationships with PCA
plot.</p>
<p><img src="images/06a_pca_analysis_cell_2d.png" width="600"></p>
<p>You can also view the plot in 3D</p>
<p><img src="images/06b_pca_analysis_dex_3d.png" width="600"></p>
<p>or with gene loadings overlaid.</p>
<p><img src="images/06c_pca_analysis_loadings.png" width="600"></p>
</li>
<li>
<p><em>MA Plot</em>: Identify differentially expressed genes with
statistical significance</p>
<p><img src="images/07_ma_plot.png" width="600"></p>
</li>
<li>
<p><em>Scatter Plot</em>: Compare fold-changes of genes across
different comparisons</p>
<p><img src="images/08_scatter.png" width="600"></p>
</li>
<li>
<p><em>Gene Plot</em>: Create customizable expression visualizations
for genes of interest</p>
<p><img src="images/09_geneplot_facet.png" width="600"></p>
</li>
<li>
<p><em>UpSet Plot</em>: Discover overlapping gene sets across
multiple comparisons</p>
<p><img src="images/10_upset.png" width="600"></p>
</li>
<li>
<p><em>Heatmap</em>: Examine expression patterns across samples and
conditions</p>
<p><img src="images/11_heatmap.png" width="600"></p>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="functional-enrichment">Functional enrichment<a class="anchor" aria-label="anchor" href="#functional-enrichment"></a>
</h4>
<p>Explore functional enrichment analysis results using this module:</p>
<ul>
<li>
<p><strong>Table</strong>: Interactive tables with powerful search
capabilities</p>
<p><img src="images/12_functable.png" width="650"></p>
</li>
<li>
<p><strong>Plots</strong>: Seven different visualizations including
network plots and dendrograms from the GeneTonic and clusterProfiler
packages:</p>
<ul>
<li><p>Summary overview to view top enriched terms <img src="images/13a_sumov.png" width="600"></p></li>
<li><p>Three flavors of enrichment map including one with similar terms
grouped together <img src="images/13b_emap_distill.png" width="600"></p></li>
<li><p>Cnetplot: network plot highlighting genes connecting functional
terms <img src="images/13c_cnetplot.png" width="600"></p></li>
<li><p>Radar plot: top enriched terms shown in a circular view <img src="images/13d_radar.png" width="600"></p></li>
<li><p>Alluvial plot: flow diagram highlighting genes connecting
functional terms <img src="images/13e_alluvial.png" width="600"></p></li>
<li><p>Dendrogram: hierarchical tree view of top enriched terms <img src="images/13f_dendrogram.png" width="600"></p></li>
</ul>
</li>
<li>
<p><strong>Compare Results</strong>: Directly compare enrichment
results between conditions using three different visualizations.</p>
<ul>
<li><p>Summary overview comparing contrasts <img src="images/14a_comp_sumov.png" width="600"></p></li>
<li>
<p>Radar plot for two comparisons</p>
<p><img src="images/14b_comp_radar.png" width="600"></p>
</li>
<li><p>Horizon plot for top terms connected by comparison <img src="images/14c_comp_horizon.png" width="600"></p></li>
</ul>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="pattern-analysis">Pattern analysis<a class="anchor" aria-label="anchor" href="#pattern-analysis"></a>
</h4>
<p>Identify co-regulated gene clusters across conditions</p>
<ul>
<li>
<strong>Plot</strong>: Visualize expression patterns of gene
clusters</li>
</ul>
<p><img src="images/15_pattern_plot.png" width="600"></p>
<ul>
<li>
<strong>Cluster Membership</strong>: Explore which genes belong to
which clusters using an interactive table</li>
</ul>
<p><img src="images/15a_pattern_tbl.png" width="600"></p>
</div>
<div class="section level4">
<h4 id="gene-scratchpad">Gene scratchpad<a class="anchor" aria-label="anchor" href="#gene-scratchpad"></a>
</h4>
<p>You can keep track of genes of interest across Carnation using the
gene scratchpad.</p>
<ul>
<li><p>To use this feature, click on the “Gene scratchpad” button in the
sidebar.</p></li>
<li>
<p>To start you can add top differentially expressed genes from a
specific comparison by using the “Top genes settings” and clicking “Add
top genes”.</p>
<p><img src="images/11a_gene_scratchpad.png" width="300"></p>
<p>Now several plots in the “DE analysis” section will have these genes
labeled, e.g. the MA plot, scatter plot, and others.</p>
<p>Labeled MA plot</p>
<p><img src="images/11b_ma_plot_label.png" width="600"></p>
<p>Labeled scatter plot</p>
<p><img src="images/11c_scatter_plot_label.png" width="600"></p>
<p>These genes can also be used to label the pattern analysis plot. Set
“label” to “gene_scratchpad” and click “Refresh plot”.</p>
<p><img src="images/11d_pattern_plt_label.png" width="400"></p>
<p>Now the genes are labeled on the plot. You will also see a table
below summarizing the clusters where these genes were found.</p>
<p><img src="images/11e_pattern_plt_label2.png" width="600"></p>
</li>
<li>
<p>You can also quickly add genes to the scratchpad from functional
enrichment tables</p>
<p>First, select your term of interest from the enrichment table</p>
<p><img src="images/11f_functbl_click.png" width="700"></p>
<p>then, click the ‘Add to scratchpad’ button at the bottom of the
page.</p>
<p><img src="images/11g_functbl_add.png" width="700"></p>
<p>Now, these genes will be added to the scratchpad and can be used to
label plots as shown above.</p>
<p><img src="images/11h_functbl_scratch_show.png" width="300"></p>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="server-mode">Server Mode<a class="anchor" aria-label="anchor" href="#server-mode"></a>
</h4>
<p>Carnation supports multi-user environments with authentication:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create user database</span></span>
<span><span class="va">credentials</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  user <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'shinymanager'</span><span class="op">)</span>,</span>
<span>  password <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'12345'</span><span class="op">)</span>,</span>
<span>  admin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Initialize the database</span></span>
<span><span class="fu">shinymanager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/shinymanager/man/create_db.html" class="external-link">create_db</a></span><span class="op">(</span></span>
<span>  credentials_data <span class="op">=</span> <span class="va">credentials</span>,</span>
<span>  sqlite_path <span class="op">=</span> <span class="st">'credentials.sqlite'</span>,</span>
<span>  passphrase <span class="op">=</span> <span class="st">'admin_passphrase'</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Now run carnation with authentication</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/run_carnation.html">run_carnation</a></span><span class="op">(</span>credentials<span class="op">=</span><span class="st">'credentials.sqlite'</span>, passphrase<span class="op">=</span><span class="st">'admin_passphrase'</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="sessioninfo">sessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">#&gt; Platform: x86_64-conda-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS/LAPACK: /home/runner/work/carnation/env/lib/libopenblasp-r0.3.30.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Etc/UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] carnation_0.99.3            DEGreport_1.42.0           </span></span>
<span><span class="co">#&gt;  [3] clusterProfiler_4.14.0      org.Hs.eg.db_3.20.0        </span></span>
<span><span class="co">#&gt;  [5] AnnotationDbi_1.68.0        GeneTonic_3.0.0            </span></span>
<span><span class="co">#&gt;  [7] dplyr_1.1.4                 DESeq2_1.46.0              </span></span>
<span><span class="co">#&gt;  [9] airway_1.26.0               SummarizedExperiment_1.36.0</span></span>
<span><span class="co">#&gt; [11] Biobase_2.66.0              GenomicRanges_1.58.0       </span></span>
<span><span class="co">#&gt; [13] GenomeInfoDb_1.42.0         IRanges_2.40.0             </span></span>
<span><span class="co">#&gt; [15] S4Vectors_0.44.0            BiocGenerics_0.52.0        </span></span>
<span><span class="co">#&gt; [17] MatrixGenerics_1.18.0       matrixStats_1.5.0          </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] fs_1.6.6                    bitops_1.0-9               </span></span>
<span><span class="co">#&gt;   [3] enrichplot_1.26.1           webshot_0.5.5              </span></span>
<span><span class="co">#&gt;   [5] httr_1.4.7                  RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;   [7] doParallel_1.0.17           dynamicTreeCut_1.63-1      </span></span>
<span><span class="co">#&gt;   [9] backports_1.5.0             tippy_0.1.0                </span></span>
<span><span class="co">#&gt;  [11] tools_4.4.3                 R6_2.6.1                   </span></span>
<span><span class="co">#&gt;  [13] DT_0.34.0                   lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt;  [15] mgcv_1.9-4                  GetoptLong_1.1.0           </span></span>
<span><span class="co">#&gt;  [17] withr_3.0.2                 prettyunits_1.2.0          </span></span>
<span><span class="co">#&gt;  [19] gridExtra_2.3               cli_3.6.5                  </span></span>
<span><span class="co">#&gt;  [21] textshaping_1.0.4           logging_0.10-108           </span></span>
<span><span class="co">#&gt;  [23] TSP_1.2.6                   labeling_0.4.3             </span></span>
<span><span class="co">#&gt;  [25] sass_0.4.10                 topGO_2.58.0               </span></span>
<span><span class="co">#&gt;  [27] bs4Dash_2.3.5               askpass_1.2.1              </span></span>
<span><span class="co">#&gt;  [29] ggridges_0.5.7              goseq_1.58.0               </span></span>
<span><span class="co">#&gt;  [31] pkgdown_2.2.0               Rsamtools_2.22.0           </span></span>
<span><span class="co">#&gt;  [33] systemfonts_1.3.1           yulab.utils_0.2.3          </span></span>
<span><span class="co">#&gt;  [35] gson_0.1.0                  txdbmaker_1.2.0            </span></span>
<span><span class="co">#&gt;  [37] DOSE_4.0.0                  R.utils_2.13.0             </span></span>
<span><span class="co">#&gt;  [39] limma_3.62.1                RSQLite_2.4.5              </span></span>
<span><span class="co">#&gt;  [41] visNetwork_2.1.4            generics_0.1.4             </span></span>
<span><span class="co">#&gt;  [43] gridGraphics_0.5-1          shape_1.4.6.1              </span></span>
<span><span class="co">#&gt;  [45] BiocIO_1.16.0               dendextend_1.19.1          </span></span>
<span><span class="co">#&gt;  [47] GO.db_3.20.0                Matrix_1.7-4               </span></span>
<span><span class="co">#&gt;  [49] abind_1.4-8                 R.methodsS3_1.8.2          </span></span>
<span><span class="co">#&gt;  [51] lifecycle_1.0.4             yaml_2.3.12                </span></span>
<span><span class="co">#&gt;  [53] edgeR_4.4.0                 qvalue_2.38.0              </span></span>
<span><span class="co">#&gt;  [55] SparseArray_1.6.0           BiocFileCache_2.14.0       </span></span>
<span><span class="co">#&gt;  [57] grid_4.4.3                  blob_1.2.4                 </span></span>
<span><span class="co">#&gt;  [59] promises_1.5.0              crayon_1.5.3               </span></span>
<span><span class="co">#&gt;  [61] miniUI_0.1.2                ggtangle_0.0.9             </span></span>
<span><span class="co">#&gt;  [63] lattice_0.22-7              billboarder_0.5.0          </span></span>
<span><span class="co">#&gt;  [65] ComplexUpset_1.3.3          cowplot_1.2.0              </span></span>
<span><span class="co">#&gt;  [67] GenomicFeatures_1.58.0      KEGGREST_1.46.0            </span></span>
<span><span class="co">#&gt;  [69] pillar_1.11.1               knitr_1.51                 </span></span>
<span><span class="co">#&gt;  [71] ComplexHeatmap_2.22.0       fgsea_1.32.2               </span></span>
<span><span class="co">#&gt;  [73] rjson_0.2.23                codetools_0.2-20           </span></span>
<span><span class="co">#&gt;  [75] fastmatch_1.1-6             glue_1.8.0                 </span></span>
<span><span class="co">#&gt;  [77] ggfun_0.2.0                 data.table_1.17.8          </span></span>
<span><span class="co">#&gt;  [79] vctrs_0.6.5                 png_0.1-8                  </span></span>
<span><span class="co">#&gt;  [81] treeio_1.30.0               gtable_0.3.6               </span></span>
<span><span class="co">#&gt;  [83] assertthat_0.2.1            cachem_1.1.0               </span></span>
<span><span class="co">#&gt;  [85] xfun_0.55                   S4Arrays_1.6.0             </span></span>
<span><span class="co">#&gt;  [87] mime_0.13                   ConsensusClusterPlus_1.70.0</span></span>
<span><span class="co">#&gt;  [89] seriation_1.5.8             shinythemes_1.2.0          </span></span>
<span><span class="co">#&gt;  [91] iterators_1.0.14            statmod_1.5.1              </span></span>
<span><span class="co">#&gt;  [93] nlme_3.1-168                ggtree_3.14.0              </span></span>
<span><span class="co">#&gt;  [95] bit64_4.6.0-1               progress_1.2.3             </span></span>
<span><span class="co">#&gt;  [97] filelock_1.0.3              rprojroot_2.1.1            </span></span>
<span><span class="co">#&gt;  [99] bslib_0.9.0                 otel_0.2.0                 </span></span>
<span><span class="co">#&gt; [101] colorspace_2.1-2            DBI_1.2.3                  </span></span>
<span><span class="co">#&gt; [103] mnormt_2.1.1                tidyselect_1.2.1           </span></span>
<span><span class="co">#&gt; [105] bit_4.6.0                   compiler_4.4.3             </span></span>
<span><span class="co">#&gt; [107] curl_7.0.0                  httr2_1.2.2                </span></span>
<span><span class="co">#&gt; [109] graph_1.84.0                BiasedUrn_2.0.12           </span></span>
<span><span class="co">#&gt; [111] SparseM_1.84-2              expm_1.0-0                 </span></span>
<span><span class="co">#&gt; [113] xml2_1.5.0                  ggdendro_0.2.0             </span></span>
<span><span class="co">#&gt; [115] desc_1.4.3                  DelayedArray_0.32.0        </span></span>
<span><span class="co">#&gt; [117] plotly_4.11.0               scrypt_0.1.6               </span></span>
<span><span class="co">#&gt; [119] colourpicker_1.3.0          rtracklayer_1.66.0         </span></span>
<span><span class="co">#&gt; [121] scales_1.4.0                psych_2.5.6                </span></span>
<span><span class="co">#&gt; [123] mosdef_1.2.0                rappdirs_0.3.3             </span></span>
<span><span class="co">#&gt; [125] stringr_1.6.0               digest_0.6.39              </span></span>
<span><span class="co">#&gt; [127] shinyBS_0.63.0              rmarkdown_2.30             </span></span>
<span><span class="co">#&gt; [129] ca_0.71.1                   XVector_0.46.0             </span></span>
<span><span class="co">#&gt; [131] htmltools_0.5.9             pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt; [133] learnr_0.11.6               dbplyr_2.5.1               </span></span>
<span><span class="co">#&gt; [135] fastmap_1.2.0               rlang_1.1.6                </span></span>
<span><span class="co">#&gt; [137] GlobalOptions_0.1.3         htmlwidgets_1.6.4          </span></span>
<span><span class="co">#&gt; [139] UCSC.utils_1.2.0            shiny_1.12.1               </span></span>
<span><span class="co">#&gt; [141] shinymanager_1.0.410        farver_2.1.2               </span></span>
<span><span class="co">#&gt; [143] jquerylib_0.1.4             jsonlite_2.0.0             </span></span>
<span><span class="co">#&gt; [145] BiocParallel_1.40.0         GOSemSim_2.32.0            </span></span>
<span><span class="co">#&gt; [147] R.oo_1.27.1                 RCurl_1.98-1.17            </span></span>
<span><span class="co">#&gt; [149] magrittr_2.0.4              GenomeInfoDbData_1.2.13    </span></span>
<span><span class="co">#&gt; [151] ggplotify_0.1.3             patchwork_1.3.2            </span></span>
<span><span class="co">#&gt; [153] Rcpp_1.1.0                  reticulate_1.44.1          </span></span>
<span><span class="co">#&gt; [155] ape_5.8-1                   shinycssloaders_1.1.0      </span></span>
<span><span class="co">#&gt; [157] viridis_0.6.5               stringi_1.8.7              </span></span>
<span><span class="co">#&gt; [159] rintrojs_0.3.4              zlibbioc_1.52.0            </span></span>
<span><span class="co">#&gt; [161] MASS_7.3-65                 plyr_1.8.9                 </span></span>
<span><span class="co">#&gt; [163] parallel_4.4.3              ggrepel_0.9.6              </span></span>
<span><span class="co">#&gt; [165] Biostrings_2.74.0           splines_4.4.3              </span></span>
<span><span class="co">#&gt; [167] hms_1.1.4                   geneLenDataBase_1.42.0     </span></span>
<span><span class="co">#&gt; [169] circlize_0.4.17             locfit_1.5-9.12            </span></span>
<span><span class="co">#&gt; [171] igraph_2.1.4                reshape2_1.4.5             </span></span>
<span><span class="co">#&gt; [173] biomaRt_2.62.0              XML_3.99-0.20              </span></span>
<span><span class="co">#&gt; [175] evaluate_1.0.5              foreach_1.5.2              </span></span>
<span><span class="co">#&gt; [177] tweenr_2.0.3                httpuv_1.6.16              </span></span>
<span><span class="co">#&gt; [179] backbone_2.1.5              openssl_2.3.4              </span></span>
<span><span class="co">#&gt; [181] tidyr_1.3.2                 purrr_1.2.0                </span></span>
<span><span class="co">#&gt; [183] polyclip_1.10-7             heatmaply_1.6.0            </span></span>
<span><span class="co">#&gt; [185] reshape_0.8.10              clue_0.3-66                </span></span>
<span><span class="co">#&gt; [187] ggplot2_3.5.2               ggforce_0.5.0              </span></span>
<span><span class="co">#&gt; [189] broom_1.0.11                xtable_1.8-4               </span></span>
<span><span class="co">#&gt; [191] restfulr_0.0.16             tidytree_0.4.6             </span></span>
<span><span class="co">#&gt; [193] later_1.4.4                 viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt; [195] ragg_1.5.0                  tibble_3.3.0               </span></span>
<span><span class="co">#&gt; [197] aplot_0.2.9                 registry_0.5-1             </span></span>
<span><span class="co">#&gt; [199] memoise_2.0.1               GenomicAlignments_1.42.0   </span></span>
<span><span class="co">#&gt; [201] cluster_2.1.8.1             sortable_0.5.0             </span></span>
<span><span class="co">#&gt; [203] shinyWidgets_0.9.0          shinyAce_0.4.4</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Apratim Mitra.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
