<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>carnation - airway tutorial • carnation</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="carnation - airway tutorial">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">carnation</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/carnation.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>carnation - airway tutorial</h1>
            
      

      <div class="d-none name"><code>carnation.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="load-libraries-airway-dataset">Load libraries &amp; airway dataset<a class="anchor" aria-label="anchor" href="#load-libraries-airway-dataset"></a>
</h3>
<p>Load carnation.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://nichd-bspc.github.io/carnation/">carnation</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span></code></pre></div>
<p>First load some libraries that we will need for this tutorial.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/federicomarini/GeneTonic" class="external-link">GeneTonic</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># install optional packages if not present</span></span>
<span><span class="va">pkgs_to_check</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'airway'</span>, <span class="st">'org.Hs.eg.db'</span>, <span class="st">'clusterProfiler'</span>, <span class="st">'DEGreport'</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">pkg</span> <span class="kw">in</span> <span class="va">pkgs_to_check</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/setRepositories.html" class="external-link">setRepositories</a></span><span class="op">(</span>ind<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">4</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="va">pkg</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="va">pkg</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pkg</span>, character.only<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>We will be using the ‘airway’ dataset. First, we load this
dataset.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">'airway'</span><span class="op">)</span></span></code></pre></div>
<p>Next, we extract the counts matrix and and metadata.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span><span class="va">cdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span></code></pre></div>
<p>Now let’s see what these look like.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 63677     8</span></span></code></pre></div>
<p>So, <code>mat</code> is a matrix with 64102 rows and 8 columns. Each
row corresponds to a single gene and each column corresponds to a single
sample. As you will notice, the rownames of <code>mat</code> contains
gene IDs and column names have the sample IDs.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 SRR1039508 SRR1039509 SRR1039512 SRR1039513 SRR1039516</span></span>
<span><span class="co">#&gt; ENSG00000000003        679        448        873        408       1138</span></span>
<span><span class="co">#&gt; ENSG00000000005          0          0          0          0          0</span></span>
<span><span class="co">#&gt; ENSG00000000419        467        515        621        365        587</span></span>
<span><span class="co">#&gt; ENSG00000000457        260        211        263        164        245</span></span>
<span><span class="co">#&gt; ENSG00000000460         60         55         40         35         78</span></span>
<span><span class="co">#&gt; ENSG00000000938          0          0          2          0          1</span></span>
<span><span class="co">#&gt;                 SRR1039517 SRR1039520 SRR1039521</span></span>
<span><span class="co">#&gt; ENSG00000000003       1047        770        572</span></span>
<span><span class="co">#&gt; ENSG00000000005          0          0          0</span></span>
<span><span class="co">#&gt; ENSG00000000419        799        417        508</span></span>
<span><span class="co">#&gt; ENSG00000000457        331        233        229</span></span>
<span><span class="co">#&gt; ENSG00000000460         63         76         60</span></span>
<span><span class="co">#&gt; ENSG00000000938          0          0          0</span></span></code></pre></div>
<p><code>cdata</code> contains the sample metadata. There is a lot of
information here, but notice the <code>cell</code> and <code>dex</code>
columns, as we will be using this for the differential expression
analysis later.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cdata</span></span>
<span><span class="co">#&gt; DataFrame with 8 rows and 9 columns</span></span>
<span><span class="co">#&gt;            SampleName     cell      dex    albut        Run avgLength</span></span>
<span><span class="co">#&gt;              &lt;factor&gt; &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;   &lt;factor&gt; &lt;integer&gt;</span></span>
<span><span class="co">#&gt; SRR1039508 GSM1275862  N61311     untrt    untrt SRR1039508       126</span></span>
<span><span class="co">#&gt; SRR1039509 GSM1275863  N61311     trt      untrt SRR1039509       126</span></span>
<span><span class="co">#&gt; SRR1039512 GSM1275866  N052611    untrt    untrt SRR1039512       126</span></span>
<span><span class="co">#&gt; SRR1039513 GSM1275867  N052611    trt      untrt SRR1039513        87</span></span>
<span><span class="co">#&gt; SRR1039516 GSM1275870  N080611    untrt    untrt SRR1039516       120</span></span>
<span><span class="co">#&gt; SRR1039517 GSM1275871  N080611    trt      untrt SRR1039517       126</span></span>
<span><span class="co">#&gt; SRR1039520 GSM1275874  N061011    untrt    untrt SRR1039520       101</span></span>
<span><span class="co">#&gt; SRR1039521 GSM1275875  N061011    trt      untrt SRR1039521        98</span></span>
<span><span class="co">#&gt;            Experiment    Sample    BioSample</span></span>
<span><span class="co">#&gt;              &lt;factor&gt;  &lt;factor&gt;     &lt;factor&gt;</span></span>
<span><span class="co">#&gt; SRR1039508  SRX384345 SRS508568 SAMN02422669</span></span>
<span><span class="co">#&gt; SRR1039509  SRX384346 SRS508567 SAMN02422675</span></span>
<span><span class="co">#&gt; SRR1039512  SRX384349 SRS508571 SAMN02422678</span></span>
<span><span class="co">#&gt; SRR1039513  SRX384350 SRS508572 SAMN02422670</span></span>
<span><span class="co">#&gt; SRR1039516  SRX384353 SRS508575 SAMN02422682</span></span>
<span><span class="co">#&gt; SRR1039517  SRX384354 SRS508576 SAMN02422673</span></span>
<span><span class="co">#&gt; SRR1039520  SRX384357 SRS508579 SAMN02422683</span></span>
<span><span class="co">#&gt; SRR1039521  SRX384358 SRS508580 SAMN02422677</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="get-more-gene-annotation">Get more gene annotation<a class="anchor" aria-label="anchor" href="#get-more-gene-annotation"></a>
</h3>
<p>The gene IDs that come with the dataset are from ENSEMBL and are not
human-readable. So, next we will extract gene symbols and
<code>ENTREZID</code> for these genes from the <code>org.Hs.eg.db</code>
package.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keytypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'SYMBOL'</span><span class="op">=</span><span class="st">'SYMBOL'</span>, <span class="st">'ENTREZID'</span><span class="op">=</span><span class="st">'ENTREZID'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">anno_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">'cbind'</span>,</span>
<span>             <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">keytypes</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>               <span class="fu">mapIds</span><span class="op">(</span><span class="va">org.Hs.eg.db</span>,</span>
<span>                 column<span class="op">=</span><span class="va">x</span>,</span>
<span>                 keys<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span>,</span>
<span>                 keytype<span class="op">=</span><span class="st">'ENSEMBL'</span><span class="op">)</span></span>
<span>               <span class="op">}</span><span class="op">)</span></span>
<span>             <span class="op">)</span></span>
<span></span>
<span><span class="co"># convert to data frame</span></span>
<span><span class="va">anno_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">anno_df</span><span class="op">)</span></span></code></pre></div>
<p>Now, we have human readable gene names in the <code>SYMBOL</code>
column and Entrez IDs in the <code>ENTREZ</code> column.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">anno_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 SYMBOL ENTREZID</span></span>
<span><span class="co">#&gt; ENSG00000000003 TSPAN6     7105</span></span>
<span><span class="co">#&gt; ENSG00000000005   TNMD    64102</span></span>
<span><span class="co">#&gt; ENSG00000000419   DPM1     8813</span></span>
<span><span class="co">#&gt; ENSG00000000457  SCYL3    57147</span></span>
<span><span class="co">#&gt; ENSG00000000460  FIRRM    55732</span></span>
<span><span class="co">#&gt; ENSG00000000938    FGR     2268</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-deseqdataset">Create DESeqDataSet<a class="anchor" aria-label="anchor" href="#create-deseqdataset"></a>
</h3>
<p>Next, we create a new <code>DESeqDataSet</code> using
<code>mat</code> and <code>cdata</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSetFromMatrix</a></span><span class="op">(</span><span class="va">mat</span>,</span>
<span>                              colData<span class="op">=</span><span class="va">cdata</span>,</span>
<span>                              design<span class="op">=</span><span class="op">~</span><span class="va">cell</span> <span class="op">+</span> <span class="va">dex</span><span class="op">)</span></span></code></pre></div>
<p>Let’s check to make sure that everything looks okay:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span></span>
<span><span class="co">#&gt; class: DESeqDataSet </span></span>
<span><span class="co">#&gt; dim: 63677 8 </span></span>
<span><span class="co">#&gt; metadata(1): version</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(63677): ENSG00000000003 ENSG00000000005 ... ENSG00000273492</span></span>
<span><span class="co">#&gt;   ENSG00000273493</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521</span></span>
<span><span class="co">#&gt; colData names(9): SampleName cell ... Sample BioSample</span></span></code></pre></div>
<p>Then we save <code>dds</code> in a list.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>main<span class="op">=</span><span class="va">dds</span><span class="op">)</span></span></code></pre></div>
<p>We also normalize the counts data and save it in a list.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rld_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">dds_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/varianceStabilizingTransformation.html" class="external-link">varianceStabilizingTransformation</a></span><span class="op">(</span><span class="va">x</span>, blind<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-differential-expression-analysis">Run differential expression analysis<a class="anchor" aria-label="anchor" href="#run-differential-expression-analysis"></a>
</h3>
<p>Now, the object <code>dds</code> is ready for differential expression
(DE) analysis.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="co">#&gt; estimating size factors</span></span>
<span><span class="co">#&gt; estimating dispersions</span></span>
<span><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span><span class="co">#&gt; mean-dispersion relationship</span></span>
<span><span class="co">#&gt; final dispersion estimates</span></span>
<span><span class="co">#&gt; fitting model and testing</span></span></code></pre></div>
<p>Since, we used <code>design=~cell + dex</code> while creating
<code>dds</code>, the above step will automatically calculate some
comparisons.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">resultsNames</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Intercept"               "cell_N061011_vs_N052611"</span></span>
<span><span class="co">#&gt; [3] "cell_N080611_vs_N052611" "cell_N61311_vs_N052611" </span></span>
<span><span class="co">#&gt; [5] "dex_untrt_vs_trt"</span></span></code></pre></div>
<p>The last comparison <code>dex_untrt_vs_trt</code> contains the effect
of the <code>dex</code> treatment, while the other comparisons compare
different cell lines. Next, we will extract two of these results and run
<code>lfcShrink</code> on them.</p>
<p>For the <code>cell</code> comparison, we choose a precomputed results
using the <code>coef</code> parameter.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_comparison</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/lfcShrink.html" class="external-link">lfcShrink</a></span><span class="op">(</span><span class="va">dds</span>,</span>
<span>                             coef<span class="op">=</span><span class="st">'cell_N61311_vs_N052611'</span>,</span>
<span>                             type<span class="op">=</span><span class="st">'normal'</span><span class="op">)</span></span></code></pre></div>
<p>For the <code>dex</code> comparison, we use <code>contrast</code> to
specify the direction of the comparison, since we want to use
<code>untrt</code> as control.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dex_comparison</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/lfcShrink.html" class="external-link">lfcShrink</a></span><span class="op">(</span><span class="va">dds</span>,</span>
<span>                            contrast<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'dex'</span>, <span class="st">'trt'</span>, <span class="st">'untrt'</span><span class="op">)</span>,</span>
<span>                            type<span class="op">=</span><span class="st">'normal'</span><span class="op">)</span></span></code></pre></div>
<p>Now, each of these comparisons, contain the DE analysis results. For
example,</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dex_comparison</span><span class="op">)</span></span>
<span><span class="co">#&gt; log2 fold change (MAP): dex trt vs untrt </span></span>
<span><span class="co">#&gt; Wald test p-value: dex trt vs untrt </span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 6 columns</span></span>
<span><span class="co">#&gt;                   baseMean log2FoldChange     lfcSE      stat      pvalue</span></span>
<span><span class="co">#&gt;                  &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSG00000000003 708.602170     -0.3741527 0.0988429 -3.787752 0.000152016</span></span>
<span><span class="co">#&gt; ENSG00000000005   0.000000             NA        NA        NA          NA</span></span>
<span><span class="co">#&gt; ENSG00000000419 520.297901      0.2020620 0.1097395  1.842943 0.065337292</span></span>
<span><span class="co">#&gt; ENSG00000000457 237.163037      0.0361672 0.1383377  0.264356 0.791505742</span></span>
<span><span class="co">#&gt; ENSG00000000460  57.932633     -0.0844567 0.2498904 -0.307054 0.758801924</span></span>
<span><span class="co">#&gt; ENSG00000000938   0.318098     -0.0841390 0.1513343 -0.393793 0.693733530</span></span>
<span><span class="co">#&gt;                       padj</span></span>
<span><span class="co">#&gt;                  &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSG00000000003 0.00128292</span></span>
<span><span class="co">#&gt; ENSG00000000005         NA</span></span>
<span><span class="co">#&gt; ENSG00000000419 0.19646985</span></span>
<span><span class="co">#&gt; ENSG00000000457 0.91141962</span></span>
<span><span class="co">#&gt; ENSG00000000460 0.89500478</span></span>
<span><span class="co">#&gt; ENSG00000000938         NA</span></span></code></pre></div>
<p>Then, we save these results in a special nested list that
<code>carnation</code> will use. Here,</p>
<ul>
<li>
<code>res</code> contains the actual DE analysis results</li>
<li>
<code>dds</code> contains the name of the DESeqDataSet used for the
DE analysis. These values should map to <code>dds_list</code> names</li>
<li>
<code>label</code> is a description of the comparison</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        dex_trt_vs_untrt<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>            res<span class="op">=</span><span class="va">dex_comparison</span>,</span>
<span>            dds<span class="op">=</span><span class="st">'main'</span>,</span>
<span>            label<span class="op">=</span><span class="st">'dex, treated vs untreated'</span><span class="op">)</span>,</span>
<span>        cell_N61311_vs_N052611<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>            res<span class="op">=</span><span class="va">cell_comparison</span>,</span>
<span>            dds<span class="op">=</span><span class="st">'main'</span>,</span>
<span>            label<span class="op">=</span><span class="st">'cell, N61311 vs N052611'</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span></code></pre></div>
<p>Finally, we add <code>SYMBOL</code> and <code>ENTREZID</code> columns
to the DE results from the <code>anno_df</code> data frame.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">res_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>              <span class="co"># save the rownames as a new 'gene' column</span></span>
<span>              <span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">[[</span> <span class="st">'gene'</span> <span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">)</span></span>
<span></span>
<span>              <span class="co"># add 'SYMBOL' and 'ENTREZID' columns</span></span>
<span>              <span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">[[</span> <span class="st">'SYMBOL'</span> <span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">anno_df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">)</span>, <span class="st">'SYMBOL'</span><span class="op">]</span></span>
<span>              <span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">[[</span> <span class="st">'ENTREZID'</span> <span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">anno_df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">)</span>, <span class="st">'ENTREZID'</span><span class="op">]</span></span>
<span></span>
<span>              <span class="va">x</span></span>
<span>            <span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-functional-enrichment-results-optional">Add functional enrichment results (optional)<a class="anchor" aria-label="anchor" href="#add-functional-enrichment-results-optional"></a>
</h3>
<p>Now we run functional enrichment on the DE genes from the two
comparisons. For this, we first set significance thresholds and then
extract the DE genes and save as a list.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># padj cutoff</span></span>
<span><span class="va">alpha</span> <span class="op">&lt;-</span> <span class="fl">0.01</span></span>
<span></span>
<span><span class="co"># log2FoldChange threshold; 1 == 2x difference</span></span>
<span><span class="va">lfc_threshold</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span></span>
<span><span class="co"># list to save DE genes</span></span>
<span><span class="va">de.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">res_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>              <span class="co"># changed genes</span></span>
<span>              <span class="va">idx</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="va">alpha</span> <span class="op">&amp;</span></span>
<span>                     <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">$</span><span class="va">padj</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>                     <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">lfc_threshold</span></span>
<span></span>
<span>              <span class="co"># return DE genes as a dataframe</span></span>
<span>              <span class="va">x</span><span class="op">$</span><span class="va">res</span><span class="op">[</span><span class="va">idx</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'gene'</span>, <span class="st">'ENTREZID'</span><span class="op">)</span><span class="op">]</span></span>
<span>            <span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>Next, we run functional enrichment and save the results in a list
called <code>enrich_list</code>. We also save a converted list called
<code>genetonic</code> which carnation uses for several plots from the
<code>GeneTonic</code> package.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># list to save functional enrichment results</span></span>
<span><span class="va">enrich_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># list to save a converted object for GeneTonic plots</span></span>
<span><span class="va">genetonic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">comp</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res_list</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">go.res</span> <span class="op">&lt;-</span> <span class="fu">clusterProfiler</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html" class="external-link">enrichGO</a></span><span class="op">(</span></span>
<span>            gene<span class="op">=</span><span class="va">de.genes</span><span class="op">[[</span><span class="va">comp</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">'ENTREZID'</span><span class="op">]</span><span class="op">]</span>,</span>
<span>            keyType<span class="op">=</span><span class="st">'ENTREZID'</span>,</span>
<span>            OrgDb<span class="op">=</span><span class="va">org.Hs.eg.db</span>,</span>
<span>            ont<span class="op">=</span><span class="st">'BP'</span>,</span>
<span>            pvalueCutoff<span class="op">=</span><span class="fl">1</span>, qvalueCutoff<span class="op">=</span><span class="fl">1</span>,</span>
<span>            readable <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">enrich_list</span><span class="op">[[</span> <span class="va">comp</span> <span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                               res<span class="op">=</span><span class="va">comp</span>,</span>
<span>                               changed<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span> BP<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">go.res</span><span class="op">)</span> <span class="op">)</span></span>
<span>                             <span class="op">)</span></span>
<span></span>
<span>    <span class="va">genetonic</span><span class="op">[[</span> <span class="va">comp</span> <span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                             res<span class="op">=</span><span class="va">comp</span>,</span>
<span>                             changed<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                               BP<span class="op">=</span><span class="fu"><a href="../reference/enrich_to_genetonic.html">enrich_to_genetonic</a></span><span class="op">(</span><span class="va">go.res</span>, <span class="va">res_list</span><span class="op">[[</span><span class="va">comp</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">res</span><span class="op">)</span></span>
<span>                             <span class="op">)</span></span>
<span>                           <span class="op">)</span></span>
<span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><code>enrich_list</code> is a nested list where:</p>
<ul>
<li>The top-level names are unique keys/identifiers</li>
<li>The second level corresponds to direction of change,
e.g. <code>up</code>, <code>down</code> or <code>changed</code>. This
level also contains a special entry <code>res</code> which maps to
<code>res_list</code> names, as a way to record where the DE results
came from.</li>
<li>The third level corresponds to the ontology, e.g. <code>BP</code>
(GO Biological Process).</li>
</ul>
<p>Here, we are just using changed genes and the GO Biological Process
(BP) ontology, so <code>enrich_list</code> looks like:</p>
<pre><code>enrich_list
  ├─ dex_trt_vs_untrt
  │    ├─ res = 'des_trt_vs_untrt'    &lt;--- comparison used to get FE results
  │    └─ changed
  │         └─ BP                     &lt;--- functional enrichment results
  │
  └─ cell_N61311_vs_N052611
       ├─ res = 'cell_N61311_vs_N052611'    &lt;--- comparison used to get FE results
       └─ changed
            └─ BP                           &lt;--- functional enrichment results
</code></pre>
<p><code>genetonic</code> mirrors the same structure.</p>
</div>
<div class="section level3">
<h3 id="add-pattern-analysis-optional">Add pattern analysis (optional)<a class="anchor" aria-label="anchor" href="#add-pattern-analysis-optional"></a>
</h3>
<p>Finally, we add some pattern analysis for the
<code>dex_trt_vs_untrt</code> comparison using the
<code>DEGreport</code> package. First, we extract normalized data for
the 755 DE genes from this comparison.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract normalized data &amp; metadata</span></span>
<span><span class="va">ma</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">rld_list</span><span class="op">[[</span><span class="st">'main'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">colData.i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rld_list</span><span class="op">[[</span><span class="st">'main'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># only keep data from DE genes</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ma</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">de.genes</span><span class="op">[[</span><span class="st">'dex_trt_vs_untrt'</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">'gene'</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">ma.i</span> <span class="op">&lt;-</span> <span class="va">ma</span><span class="op">[</span><span class="va">idx</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># remove any genes with 0 variance</span></span>
<span><span class="va">ma.i</span> <span class="op">&lt;-</span> <span class="va">ma.i</span><span class="op">[</span><span class="fu">rowVars</span><span class="op">(</span><span class="va">ma.i</span><span class="op">)</span> <span class="op">!=</span> <span class="fl">0</span>, <span class="op">]</span></span></code></pre></div>
<p>Then, we run the pattern analysis, using <code>cell</code> as the
<em>time</em> variable and <code>dex</code> as the <em>color</em>
variable.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">DEGreport</span><span class="fu">::</span><span class="fu"><a href="http://lpantano.github.io/DEGreport/reference/degPatterns.html" class="external-link">degPatterns</a></span><span class="op">(</span></span>
<span>        <span class="va">ma.i</span>,</span>
<span>        <span class="va">colData.i</span>,</span>
<span></span>
<span>        time<span class="op">=</span><span class="st">'cell'</span>,</span>
<span>        col<span class="op">=</span><span class="st">'dex'</span>,</span>
<span></span>
<span>        <span class="co"># NOTE: reduce and merge cutoff----------------------------------------</span></span>
<span>        <span class="co">#   Reduce will merge clusters that are similar; similarity determined</span></span>
<span>        <span class="co">#   by cutoff</span></span>
<span>        reduce<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span></span>
<span>        <span class="co"># NOTE: set minimum cluster size to 1</span></span>
<span>        minc<span class="op">=</span><span class="fl">1</span>,</span>
<span>        plot<span class="op">=</span><span class="cn">FALSE</span></span>
<span>        <span class="op">)</span></span>
<span><span class="co">#&gt; Working with 754 genes.</span></span>
<span><span class="co">#&gt; Working with 650 genes after filtering: minc &gt; 1</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(merge)`</span></span>
<span><span class="co">#&gt; Joining with `by = join_by(merge)`</span></span></code></pre></div>
<p>Next, we extract the <code>normalized</code> slot from this object
and save as a list.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract normalized slot and add symbol column</span></span>
<span><span class="va">p_norm</span> <span class="op">&lt;-</span> <span class="va">p</span><span class="op">$</span><span class="va">normalized</span></span>
<span><span class="va">p_norm</span><span class="op">[[</span> <span class="st">'SYMBOL'</span> <span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">anno_df</span><span class="op">[</span><span class="va">p_norm</span><span class="op">[[</span><span class="st">'genes'</span><span class="op">]</span><span class="op">]</span>, <span class="st">'SYMBOL'</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># save pattern analysis results</span></span>
<span><span class="va">degpatterns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>dex_by_cell<span class="op">=</span><span class="va">p_norm</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="compose-carnation-object">Compose carnation object<a class="anchor" aria-label="anchor" href="#compose-carnation-object"></a>
</h3>
<p>Now we have all the pieces to build the carnation object.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>res_list<span class="op">=</span><span class="va">res_list</span>,</span>
<span>                 dds_list<span class="op">=</span><span class="va">dds_list</span>,</span>
<span>                 rld_list<span class="op">=</span><span class="va">rld_list</span>,</span>
<span>                 enrich_list<span class="op">=</span><span class="va">enrich_list</span>,</span>
<span>                 genetonic<span class="op">=</span><span class="va">genetonic</span>,</span>
<span>                 degpatterns_list<span class="op">=</span><span class="va">degpatterns</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">combined</span>, <span class="st">'carnation_vignette.rds'</span>, compress<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="sessioninfo">sessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] DEGreport_1.42.0            clusterProfiler_4.14.6     </span></span>
<span><span class="co">#&gt;  [3] org.Hs.eg.db_3.20.0         AnnotationDbi_1.68.0       </span></span>
<span><span class="co">#&gt;  [5] airway_1.26.0               GeneTonic_3.0.0            </span></span>
<span><span class="co">#&gt;  [7] dplyr_1.1.4                 DESeq2_1.46.0              </span></span>
<span><span class="co">#&gt;  [9] SummarizedExperiment_1.36.0 Biobase_2.66.0             </span></span>
<span><span class="co">#&gt; [11] MatrixGenerics_1.18.1       matrixStats_1.5.0          </span></span>
<span><span class="co">#&gt; [13] GenomicRanges_1.58.0        GenomeInfoDb_1.42.3        </span></span>
<span><span class="co">#&gt; [15] IRanges_2.40.1              S4Vectors_0.44.0           </span></span>
<span><span class="co">#&gt; [17] BiocGenerics_0.52.0         carnation_1.1              </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] fs_1.6.5                    bitops_1.0-9               </span></span>
<span><span class="co">#&gt;   [3] enrichplot_1.26.6           httr_1.4.7                 </span></span>
<span><span class="co">#&gt;   [5] webshot_0.5.5               RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;   [7] doParallel_1.0.17           dynamicTreeCut_1.63-1      </span></span>
<span><span class="co">#&gt;   [9] backports_1.5.0             tippy_0.1.0                </span></span>
<span><span class="co">#&gt;  [11] tools_4.4.3                 R6_2.6.1                   </span></span>
<span><span class="co">#&gt;  [13] DT_0.33                     lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt;  [15] mgcv_1.9-1                  GetoptLong_1.0.5           </span></span>
<span><span class="co">#&gt;  [17] withr_3.0.2                 prettyunits_1.2.0          </span></span>
<span><span class="co">#&gt;  [19] gridExtra_2.3               cli_3.6.4                  </span></span>
<span><span class="co">#&gt;  [21] textshaping_1.0.0           logging_0.10-108           </span></span>
<span><span class="co">#&gt;  [23] TSP_1.2-4                   labeling_0.4.3             </span></span>
<span><span class="co">#&gt;  [25] sass_0.4.9                  topGO_2.58.0               </span></span>
<span><span class="co">#&gt;  [27] bs4Dash_2.3.4               askpass_1.2.1              </span></span>
<span><span class="co">#&gt;  [29] ggridges_0.5.6              goseq_1.58.0               </span></span>
<span><span class="co">#&gt;  [31] pkgdown_2.1.1               Rsamtools_2.22.0           </span></span>
<span><span class="co">#&gt;  [33] systemfonts_1.2.1           yulab.utils_0.2.0          </span></span>
<span><span class="co">#&gt;  [35] gson_0.1.0                  txdbmaker_1.2.1            </span></span>
<span><span class="co">#&gt;  [37] DOSE_4.0.0                  R.utils_2.13.0             </span></span>
<span><span class="co">#&gt;  [39] limma_3.62.2                RSQLite_2.3.9              </span></span>
<span><span class="co">#&gt;  [41] visNetwork_2.1.2            generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [43] gridGraphics_0.5-1          shape_1.4.6.1              </span></span>
<span><span class="co">#&gt;  [45] BiocIO_1.16.0               dendextend_1.19.0          </span></span>
<span><span class="co">#&gt;  [47] GO.db_3.20.0                Matrix_1.7-2               </span></span>
<span><span class="co">#&gt;  [49] abind_1.4-8                 R.methodsS3_1.8.2          </span></span>
<span><span class="co">#&gt;  [51] lifecycle_1.0.4             edgeR_4.4.2                </span></span>
<span><span class="co">#&gt;  [53] yaml_2.3.10                 qvalue_2.38.0              </span></span>
<span><span class="co">#&gt;  [55] SparseArray_1.6.2           BiocFileCache_2.14.0       </span></span>
<span><span class="co">#&gt;  [57] grid_4.4.3                  blob_1.2.4                 </span></span>
<span><span class="co">#&gt;  [59] promises_1.3.2              crayon_1.5.3               </span></span>
<span><span class="co">#&gt;  [61] miniUI_0.1.1.1              ggtangle_0.0.6             </span></span>
<span><span class="co">#&gt;  [63] lattice_0.22-6              billboarder_0.5.0          </span></span>
<span><span class="co">#&gt;  [65] ComplexUpset_1.3.3          cowplot_1.1.3              </span></span>
<span><span class="co">#&gt;  [67] GenomicFeatures_1.58.0      KEGGREST_1.46.0            </span></span>
<span><span class="co">#&gt;  [69] pillar_1.10.1               knitr_1.49                 </span></span>
<span><span class="co">#&gt;  [71] ComplexHeatmap_2.22.0       fgsea_1.32.2               </span></span>
<span><span class="co">#&gt;  [73] rjson_0.2.23                codetools_0.2-20           </span></span>
<span><span class="co">#&gt;  [75] fastmatch_1.1-6             glue_1.8.0                 </span></span>
<span><span class="co">#&gt;  [77] ggfun_0.1.8                 data.table_1.17.0          </span></span>
<span><span class="co">#&gt;  [79] vctrs_0.6.5                 png_0.1-8                  </span></span>
<span><span class="co">#&gt;  [81] treeio_1.30.0               gtable_0.3.6               </span></span>
<span><span class="co">#&gt;  [83] assertthat_0.2.1            cachem_1.1.0               </span></span>
<span><span class="co">#&gt;  [85] xfun_0.51                   S4Arrays_1.6.0             </span></span>
<span><span class="co">#&gt;  [87] mime_0.12                   ConsensusClusterPlus_1.70.0</span></span>
<span><span class="co">#&gt;  [89] seriation_1.5.7             shinythemes_1.2.0          </span></span>
<span><span class="co">#&gt;  [91] iterators_1.0.14            statmod_1.5.0              </span></span>
<span><span class="co">#&gt;  [93] nlme_3.1-167                ggtree_3.14.0              </span></span>
<span><span class="co">#&gt;  [95] bit64_4.6.0-1               progress_1.2.3             </span></span>
<span><span class="co">#&gt;  [97] filelock_1.0.3              rprojroot_2.0.4            </span></span>
<span><span class="co">#&gt;  [99] bslib_0.9.0                 colorspace_2.1-1           </span></span>
<span><span class="co">#&gt; [101] DBI_1.2.3                   mnormt_2.1.1               </span></span>
<span><span class="co">#&gt; [103] tidyselect_1.2.1            bit_4.6.0                  </span></span>
<span><span class="co">#&gt; [105] compiler_4.4.3              curl_6.2.1                 </span></span>
<span><span class="co">#&gt; [107] httr2_1.1.1                 graph_1.84.1               </span></span>
<span><span class="co">#&gt; [109] BiasedUrn_2.0.12            SparseM_1.84-2             </span></span>
<span><span class="co">#&gt; [111] expm_1.0-0                  xml2_1.3.7                 </span></span>
<span><span class="co">#&gt; [113] ggdendro_0.2.0              desc_1.4.3                 </span></span>
<span><span class="co">#&gt; [115] DelayedArray_0.32.0         plotly_4.10.4              </span></span>
<span><span class="co">#&gt; [117] scrypt_0.1.6                colourpicker_1.3.0         </span></span>
<span><span class="co">#&gt; [119] rtracklayer_1.66.0          scales_1.3.0               </span></span>
<span><span class="co">#&gt; [121] psych_2.4.12                mosdef_1.2.0               </span></span>
<span><span class="co">#&gt; [123] rappdirs_0.3.3              stringr_1.5.1              </span></span>
<span><span class="co">#&gt; [125] digest_0.6.37               shinyBS_0.61.1             </span></span>
<span><span class="co">#&gt; [127] rmarkdown_2.29              ca_0.71.1                  </span></span>
<span><span class="co">#&gt; [129] XVector_0.46.0              htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt; [131] pkgconfig_2.0.3             learnr_0.11.5              </span></span>
<span><span class="co">#&gt; [133] dbplyr_2.5.0                fastmap_1.2.0              </span></span>
<span><span class="co">#&gt; [135] rlang_1.1.5                 GlobalOptions_0.1.2        </span></span>
<span><span class="co">#&gt; [137] htmlwidgets_1.6.4           UCSC.utils_1.2.0           </span></span>
<span><span class="co">#&gt; [139] shiny_1.10.0                farver_2.1.2               </span></span>
<span><span class="co">#&gt; [141] jquerylib_0.1.4             shinymanager_1.0.410       </span></span>
<span><span class="co">#&gt; [143] jsonlite_1.9.1              BiocParallel_1.40.0        </span></span>
<span><span class="co">#&gt; [145] GOSemSim_2.32.0             R.oo_1.27.0                </span></span>
<span><span class="co">#&gt; [147] RCurl_1.98-1.16             magrittr_2.0.3             </span></span>
<span><span class="co">#&gt; [149] GenomeInfoDbData_1.2.13     ggplotify_0.1.2            </span></span>
<span><span class="co">#&gt; [151] patchwork_1.3.0             munsell_0.5.1              </span></span>
<span><span class="co">#&gt; [153] Rcpp_1.0.14                 ape_5.8-1                  </span></span>
<span><span class="co">#&gt; [155] shinycssloaders_1.1.0       viridis_0.6.5              </span></span>
<span><span class="co">#&gt; [157] reticulate_1.41.0.1         stringi_1.8.4              </span></span>
<span><span class="co">#&gt; [159] rintrojs_0.3.4              zlibbioc_1.52.0            </span></span>
<span><span class="co">#&gt; [161] MASS_7.3-64                 plyr_1.8.9                 </span></span>
<span><span class="co">#&gt; [163] parallel_4.4.3              ggrepel_0.9.6              </span></span>
<span><span class="co">#&gt; [165] Biostrings_2.74.1           splines_4.4.3              </span></span>
<span><span class="co">#&gt; [167] hms_1.1.3                   geneLenDataBase_1.42.0     </span></span>
<span><span class="co">#&gt; [169] circlize_0.4.16             locfit_1.5-9.12            </span></span>
<span><span class="co">#&gt; [171] igraph_2.1.4                reshape2_1.4.4             </span></span>
<span><span class="co">#&gt; [173] biomaRt_2.62.1              XML_3.99-0.18              </span></span>
<span><span class="co">#&gt; [175] evaluate_1.0.3              foreach_1.5.2              </span></span>
<span><span class="co">#&gt; [177] tweenr_2.0.3                httpuv_1.6.15              </span></span>
<span><span class="co">#&gt; [179] backbone_2.1.4              openssl_2.3.2              </span></span>
<span><span class="co">#&gt; [181] tidyr_1.3.1                 purrr_1.0.4                </span></span>
<span><span class="co">#&gt; [183] polyclip_1.10-7             reshape_0.8.9              </span></span>
<span><span class="co">#&gt; [185] heatmaply_1.5.0             clue_0.3-66                </span></span>
<span><span class="co">#&gt; [187] ggplot2_3.5.1               ggforce_0.4.2              </span></span>
<span><span class="co">#&gt; [189] broom_1.0.7                 xtable_1.8-4               </span></span>
<span><span class="co">#&gt; [191] restfulr_0.0.15             tidytree_0.4.6             </span></span>
<span><span class="co">#&gt; [193] later_1.4.1                 viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt; [195] ragg_1.3.3                  tibble_3.2.1               </span></span>
<span><span class="co">#&gt; [197] aplot_0.2.5                 memoise_2.0.1              </span></span>
<span><span class="co">#&gt; [199] registry_0.5-1              GenomicAlignments_1.42.0   </span></span>
<span><span class="co">#&gt; [201] cluster_2.1.8               sortable_0.5.0             </span></span>
<span><span class="co">#&gt; [203] shinyWidgets_0.9.0          shinyAce_0.4.4</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Apratim Mitra.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
